src/upload.ts

```diff
@@ -0,0 +1,150 @@
+    }
+  };
+  uploadInfos[sequenceNo] = { compressed, raw, failureCount: 0 };
+  uploadHandler(compressed, onSuccess, onFailure);
+
+  if (config.debug) {
+    let rawLength = JSON.stringify(raw).length;
+    debug(`** Upload #${uploadCount}: Batch #${sequenceNo}`
+          + ` ${Math.round(compressed.length / 1024.0)}KB`
+          + ` (${Math.round(rawLength / 1024.0)}KB Raw). **`);
+  }
+  uploadCount++;
+}
+
+export function enqueueUpload(compressed: string, raw: IPayload) {
+  let payloadInfo: IUploadInfo = { compressed, raw, failureCount: 0 };
+  uploadQueue.push(payloadInfo);
+}
+
+export function flushUploadQueue() {
+  uploadSequentially(uploadQueue);
+}
+
+export function uploadSequentially(uploads: IUploadInfo[]) {
+  if (uploads.length > 0) {
+    let nextUpload = uploads.shift();
+    let uploadNext = (status: number) => {
+      uploadSequentially(uploads);
+    };
+    upload(nextUpload.compressed, nextUpload.raw, uploadNext, uploadNext);
+  }
+}
+
+export function resetUploads() {
+  sentBytesCount = 0;
+  uploadCount = 0;
+  uploadQueue = [];
+  uploadInfos = {};
+  reUploadQueue = [];
+}
+
+function defaultUpload(payload: string, onSuccess?: UploadCallback, onFailure?: UploadCallback) {
+  if (config.uploadUrl.length > 0) {
+    payload = JSON.stringify(payload);
+    let xhr = new XMLHttpRequest();
+    xhr.open("POST", config.uploadUrl);
+    xhr.setRequestHeader("Content-Type", "application/json");
+    xhr.onreadystatechange = () => { onXhrReadyStatusChange(xhr, onSuccess, onFailure); };
+    xhr.send(payload);
+  }
+}
+
+function onXhrReadyStatusChange(xhr: XMLHttpRequest, onSuccess: UploadCallback, onFailure: UploadCallback) {
+  if (xhr.readyState === XMLHttpRequest.DONE) {
+    // HTTP response status documentation:
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
+    if (xhr.status < 200 || xhr.status > 208) {
+      onFailure(xhr.status);
+    } else {
+      onSuccess(xhr.status);
+    }
+  }
+}
+
+function onSuccessDefault(sequenceNumber: number, status: number) {
+  debug(`SUCCESS: Delivered batch #${sequenceNumber}`);
+  let sentData = uploadInfos[sequenceNumber];
+  delete uploadInfos[sequenceNumber];
+  sentBytesCount += sentData.compressed.length;
+
+  if (state === State.Activated && sentBytesCount > config.totalLimit) {
+    let totalByteLimitExceededEventState: ITotalByteLimitExceededEventState = {
+      type: Instrumentation.TotalByteLimitExceeded,
+      bytes: sentBytesCount
+    };
+    instrument(totalByteLimitExceededEventState);
+    teardown();
+    return;
+  }
+
+  reUpload();
+}
+
+function onFailureDefault(sequenceNumber: number, status: number) {
+  debug(`FAILED: Delivery failed for batch #${sequenceNumber} Status: ${status}`);
+  let uploadInfo = uploadInfos[sequenceNumber];
+  uploadInfo.failureCount++;
+  logXhrError(sequenceNumber, status, uploadInfo);
+  if (uploadInfo.failureCount <= config.reUploadLimit) {
+    reUploadQueue.push(sequenceNumber);
+  }
```

Comment:
I don't think it makes sense to continue uploading if we exceed the failure limit, best case scenario we are going to have a hole in the middle of the impression and a playback that will either not render or will be messed up

Reply:
I had similar thoughts, but concluded there's still merit for analysis. We might still be able to extract things like the length of impression, instrumentation events, error data etc even if visualization is broken.

---
In reply to: [168000764](https://github.com/Microsoft/clarity-js/pull/63#discussion_r168000764) [](ancestors = 168000764)